hcl/fmt.go

```diff
@@ -151,3 +164,318 @@ func (f FormatResult) Formatted() string {
 const (
 	errFormatTree errors.Kind = "formatting tree"
 )
+
+func fmtBody(body *hclwrite.Body) {
+	logger := log.With().
+		Str("action", "hcl.fmtBody()").
+		Logger()
+
+	attrs := body.Attributes()
+	for name, attr := range attrs {
+		logger.Trace().
+			Str("name", name).
+			Msg("formatting attribute")
+		body.SetAttributeRaw(name, fmtAttrExpr(attr.Expr().BuildTokens(nil)))
+	}
+
+	blocks := body.Blocks()
+	for _, block := range blocks {
+		fmtBody(block.Body())
+	}
+}
+
+func fmtAttrExpr(tokens hclwrite.Tokens) hclwrite.Tokens {
+	logger := log.With().
+		Str("action", "hcl.fmtAttrExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("trimming newlines")
+
+	trimmed := trimNewlines(tokens)
+
+	logger = logger.With().
+		Str("trimmedTokens", tokensStr(trimmed)).
+		Logger()
+
+	// We are interested on lists, ignore the rest
+	if trimmed[0].Type == hclsyntax.TokenOBrack {
+
+		if isListComprehension(tokens) {
+			logger.Trace().Msg("list comprehension, ignoring")
+			return tokens
+		}
+
+		formattedList, pos := fmtListExpr(trimmed)
+		if pos != len(trimmed) {
+			panic(fmt.Errorf(
+				"last pos %d != tokens len %d for tokens: %q",
+				pos,
+				len(trimmed),
+				tokensStr(trimmed),
+			))
+		}
+		return formattedList
+	}
+
+	logger.Trace().Msg("not a list, returning tokens as is")
+	return trimmed
+}
+
+// fmtListExpr will adjust the given list tokens so they can be formatted
+// properly. It returns the adjusted tokens and the position of the first
+// token after the list ended.
+//
+// If there is no more tokens after the end of
+// the list the returned position will be equal to len(tokens).
+func fmtListExpr(tokens hclwrite.Tokens) (hclwrite.Tokens, int) {
+	logger := log.With().
+		Str("action", "hcl.fmtListExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("formatting list")
+
+	elemNextPos := 0
+	newTokens := hclwrite.Tokens{tokens[elemNextPos], newlineToken()}
+	elemNextPos++
+
+	skipNls := func() {
+		_, skipped := skipNewlines(tokens[elemNextPos:])
+		elemNextPos += skipped
+	}
+
+	for {
+		skipNls()
+
+		nextTokenType := tokens[elemNextPos].Type
+
+		if nextTokenType == hclsyntax.TokenComma {
```

Comment:
I think it could be in the end, I just didn't shared the notion of it being odd the way it is. In my head I was always extracting each element (and the element only), so it is fair to assume that when a new loop start you may be at the end of the list `]` or there may be more elements `,`. So in my head it kinda made sense/was OK :thinking: . But if for more seasoned parsing people this is terrible I can try to move it to the end, should be possible.

Reply:
ah true, sorry I have the mindset of "crazy motherfucker parser that is not a parser actually" :laughing: